walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+      return allAvailable;
+    }
+
+    // Try finding time with all attendees, if none can be found try finding time with
+    // only mandatory attendees.
+    Collection<TimeRange> times = findPotentialTimes(allBookedEvents, request);
+    if(times.isEmpty()){
+      return findPotentialTimes(mandatoryBookedEvents, request);
+    }
+    return times;
   }
+
+  /**
+   * Finds all the events that attendees in the request are booked for.
+   * @param events - all possible events.
+   * @param attendees - from the meeting request.
+   * @return events that request attendees are booked for.
+   */
+  public List<Event> findBookedEvents(Collection<Event> events, Collection<String> attendees) {
+    // List of existing bookedEvents for attendees in meeting request.
+    List<Event> bookedEvents = new ArrayList<>();
+
+    // Find all events that attendees in the request are already booked for.
+    for(Event event : events){
+      for (String attendee : attendees){
+        if(event.getAttendees().contains(attendee) && !bookedEvents.contains(event)){
+          bookedEvents.add(event);
+        }
+      }
+    }
+    return bookedEvents;
+  } 
+  /**
+   * Finds all potential availabilities in between bookedEvents.
+   * @param bookedEvents - timerange already booked out/unavailable.
+   * @param request - meetin request to add in a potential new timerange.
+   * @return list of timeranges for availability.
+   */
+  public Collection<TimeRange> findPotentialTimes(List<Event> bookedEvents, MeetingRequest request){
+    // List of potential free timeranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+    
+    Event prevEvent = bookedEvents.get(0); 
+    Event currEvent = bookedEvents.get(0);
+    for(int i = 0; i < bookedEvents.size(); i++){ 
+      prevEvent = currEvent; 
+      currEvent = bookedEvents.get(i); 
+
+      // If only one event - consider only timerange before and after the event.
+      if(bookedEvents.size() == 1) {
+        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
+        potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        break;
+      }
+
+      // No times found yet.
+      if(potentialTime.size() == 0) { 
+        // See if there is a slot before the first booked event and if this slot.
+        // is large enough for the request. 
+        if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+          // If request duration is within slot between start of day and first event (curr).
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
```

Comment:
rather than adding and removing this end of day timerange for every loop it might be easier to just do a check after the for loop for everything from the last meeting to the end of the day.
That way you could probably remove a bunch of the special cases in here too, you just need to check between the current and previous event every time (with a special case for i = 0) and then have a final if statement after hte for loop.

Reply:
Done